EmbeddedSurfacesParallelSynchronization.cpp

```diff
@@ -0,0 +1,459 @@
+  // We need to send over the new embedded surfaces and related objects for those whose parents are ghosts on neighbors.
+
+  MPI_iCommData commData;
+  commData.resize( neighbors.size());
+  for( unsigned int neighborIndex=0; neighborIndex<neighbors.size(); ++neighborIndex )
+  {
+    NeighborCommunicator & neighbor = neighbors[neighborIndex];
+
+    packNewObjectsToGhosts( &neighbor,
+                            commData.commID,
+                            mesh,
+                            newObjects );
+
+    neighbor.mpiISendReceiveBufferSizes( commData.commID,
+                                         commData.mpiSizeSendBufferRequest[neighborIndex],
+                                         commData.mpiSizeRecvBufferRequest[neighborIndex],
+                                         MPI_COMM_GEOSX );
+
+  }
+
+  for( unsigned int count=0; count<neighbors.size(); ++count )
+  {
+    int neighborIndex;
+    MpiWrapper::waitAny( commData.size,
+                         commData.mpiSizeRecvBufferRequest.data(),
+                         &neighborIndex,
+                         commData.mpiSizeRecvBufferStatus.data() );
+
+    NeighborCommunicator & neighbor = neighbors[neighborIndex];
+
+    neighbor.mpiISendReceiveBuffers( commData.commID,
+                                     commData.mpiSendBufferRequest[neighborIndex],
+                                     commData.mpiRecvBufferRequest[neighborIndex],
+                                     MPI_COMM_GEOSX );
+  }
+
+
+  for( unsigned int count=0; count<neighbors.size(); ++count )
+  {
+
+    int neighborIndex = count;
+    if( mpiCommOrder == 0 )
+    {
+      MpiWrapper::waitAny( commData.size,
+                           commData.mpiRecvBufferRequest.data(),
+                           &neighborIndex,
+                           commData.mpiRecvBufferStatus.data() );
+    }
+    else
+    {
+      MpiWrapper::wait( commData.mpiRecvBufferRequest.data() + count,
+                        commData.mpiRecvBufferStatus.data() + count );
+    }
+
+    NeighborCommunicator & neighbor = neighbors[neighborIndex];
+
+    unpackNewToGhosts( &neighbor, commData.commID, mesh );
+  }
+}
+
+void EmebeddedSurfacesParallelSynchronization::packNewObjectsToGhosts( NeighborCommunicator * const neighbor,
+                                                                       int commID,
+                                                                       MeshLevel & mesh,
+                                                                       NewObjectLists & newObjects )
+{
+  EmbeddedSurfaceNodeManager & nodeManager = mesh.getEmbSurfNodeManager();
+  EdgeManager const & edgeManager = mesh.getEdgeManager();
+  ElementRegionManager & elemManager = mesh.getElemManager();
+
+  int neighborRank = neighbor->neighborRank();
+
+  localIndex_array newNodesToSend;
+  map< std::pair< localIndex, localIndex >, std::set< localIndex > > newSurfaceGhostsToSend;
```

Comment:
Again it should be an `unordered_map` if the ordering isn't required.

Reply:
I think this can be unordered but I get an error which I cannot understand.

```
/Library/Developer/CommandLineTools/usr/include/c++/v1/__hash_table:868:5: error: static_assert failed due to requirement
      '__check_hash_requirements<pair<long, long>, hash<pair<long, long> > >::value' "the specified hash does not meet the Hash requirements"
    static_assert(__check_hash_requirements<_Key, _Hash>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__hash_table:883:1: note: in instantiation of template class
      'std::__1::__enforce_unordered_container_requirements<std::__1::pair<long, long>, std::__1::hash<std::__1::pair<long, long> >,
      std::__1::equal_to<std::__1::pair<long, long> > >' requested here
typename __enforce_unordered_container_requirements<_Key, _Hash, _Equal>::type

Reply:
https://stackoverflow.com/questions/32685540/why-cant-i-compile-an-unordered-map-with-a-pair-as-key looks like you cannot just use a pair in an `unordered_map`

Reply:
Wow that's lame the standard doesn't provide a hash for `std::pair`. Well we can add one later and change this and others.`std::unordered_map` can be orders of magnitude faster than `std::map`.